Angular: Improve decorators

ThibaudAV opened the pull request:
Issue: #12749 #10272

I think this solves this issue:
https://github.com/storybookjs/storybook/issues/12749

And gives a solution for this one
https://github.com/storybookjs/storybook/issues/10272 

Remove this note in decorators documentation :
> // Note: To see the global decorator applied to your stories you'll need to include the template key with the component's selector with each story you have.


## What I did

- Transforms the component (if present) into a template so that it can be included in other templates. 
- Add of 2 decorators function for ease of use


<details><summary>Exemple before/after</summary>
Before: 

```js
export default {
  title: 'Example/Button',
  component: ButtonComponent,
  decorators: [
    (storyFunc) => {
      const story = storyFunc();

      return {
        ...story,
        template: `<div style="margin: 3em">${story.template}</div>`,
      };
    },
  ],
} as Meta;

// Note:  To see the decorator applied to your component's stories you'll need to include the template key with the component's selector.

const Template: Story<ButtonComponent> = (args: ButtonComponent) => ({
  component: ButtonComponent,
  moduleMetadata: {
    declarations: [ButtonComponent],
  },
  props: args,
  template: '<app-button label="Submit"></app-button>',
});
```

After :
```js
export default {
  title: 'Example/Button',
  component: ButtonComponent,
  decorators: [
    moduleMetadata({
      declarations: [ParentComponent],
    }),
    // With template
    componentWrapperDecorator((stroy) => `<div style="margin: 3em">${story}</div>`),
    // With component which contains ng-content
    componentWrapperDecorator(ParentComponent),
  ],
} as Meta;

const Template: Story<ButtonComponent> = (args: ButtonComponent) => ({
  props: args,
  template: '<app-button label="Submit"></app-button>',
});

// !! this was not possible before 
const Component: Story<ButtonComponent> = (args: ButtonComponent) => ({
  props: args,
});
```

</details>


<details><summary>Exemple with ng-content</summary>

```js 
// ParentComponent
@Component({
  selector: 'parent-component',
  template: `<div style="margin: 3em; border:solid;"><ng-content></ng-content></div>`,
})
export default class ParentComponent {}

// Story 

export default {
  title: 'Example/Button',
  /**
To get the docs of parent component.
But needs to add a template in the Story otherwise storybook will take this component here by default 
We can also put child component according to story goal
*/
  component: ParentComponent, 
  decorators: [
    componentWrapperDecorator(ParentComponent),
  ],
} as Meta;

const EmptyContent = (args) => ({
  props: args,
  template: '',
});

const WithContent = (args) => ({
  props: args,
  template: 'The content',
});


```
</details>


## How to test

- Is this testable with Jest or Chromatic screenshots? na
- Does this need a new example in the kitchen sink apps? na
- Does this need an update to the documentation? yes

-> Go on each story to check them 
-> Thanks to the storyshots tests it is clear that the behaviour does not change 🤞 

## TODO

- [ ] get feedback 🙏 
- [x] (TODO in another PR) ~~Improve moduleMetadata decorator type~~
- [x] (TODO in another PR) ~~move preview/angular/decorators.ts in /preview/decorators.ts~~
- [x] add props (with args) in decorator function

ThibaudAV committed:
docs(angular): add required title in default export

ThibaudAV committed:
refactor(angular): separates StorybookModuleMetadata from rendering service

and rename some function for more clarity

ThibaudAV committed:
feat(angular): improve Decorators and Angular template generation

ThibaudAV commented on app/angular/src/client/preview/angular/decorators.ts:
```diff
@@ -20,3 +23,20 @@ export const moduleMetadata = (metadata: Partial<NgModuleMetadata>) => (storyFn:
   };
 };
+
+export const componentDecorator = (
```
we could merge the 2 `componentDecorator` and `templateDecorator` 🤷‍♂️ 

Do not hesitate to suggest a better name 🙏

gaetanmaisse replied:
Are there any benefits of keeping these 2 decorators separated? As a user, I think it would be much more understandable if we get a single `componentWrapperDecorator` that can take either a template or a Component as 1st param. What do you think?

ThibaudAV replied:
> Are there any benefits of keeping these 2 decorators separated?
I don't see any. 🚀  I'm merging them together

shilman commented on examples/angular-cli/src/stories/decorators/decorators.stories.ts:
```diff
@@ -0,0 +1,47 @@
+  component: ChildComponent,
+  decorators: [
+    templateDecorator(
+      (stroy) => `Grandparent<br><div style="margin: 3em; border:solid;">${stroy}</div>`
```
```suggestion
      (story) => `Grandparent<br><div style="margin: 3em; border:solid;">${story}</div>`
```

shilman commented on app/angular/src/client/preview/angular-beta/StorybookWrapperComponent.ts:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    ngAfterViewInit(): void {
+      // Bind properties to component, if the stroy have component
```
```suggestion
      // Bind properties to component, if the story have component
```

shilman commented on app/angular/src/client/preview/angular/decorators.ts:
```diff
@@ -20,3 +23,20 @@ export const moduleMetadata = (metadata: Partial<NgModuleMetadata>) => (storyFn:
+};
+
+export const templateDecorator = (
+  template: (stroy: string) => string
```
```suggestion
  template: (story: string) => string
```

shilman reviewed (commented):
Hi @ThibaudAV, I fixed a few typos here, but overall it looks like an improvement to me--I think we have some similar syntactic sugar for decorators in Vue.

One question: is this a breaking change? It seems like the type of `story` is now `string`, but is this only for `angular-beta` users? If so, do we need some kind of `MIGRATION.md` documentation for users migrating from `angular` to `angular-beta`?

shilman reviewed (commented):
@kroeder @Marklb can you please take a look when you have time?

ThibaudAV committed:
docs(angular): update examples with `componentWrapperDecorator`

ThibaudAV committed:
fix(angular): reset route element when a new rendering is made

allow beta and legacy redering mode to work together

ThibaudAV committed:
feat(angular): add more exemple in decorators.stories.ts

ThibaudAV commented:
> One question: is this a breaking change? It seems like the type of story is now string, but is this only for angular-beta users? If so, do we need some kind of MIGRATION.md documentation for users migrating from angular to angular-beta?

I don't think there is a breaking change. The way without syntactic sugar works as before.
no the story is not a string now (if I'm not mistaken). (I think this could be a dedicated subject)
I have tested only with angular-beta. for the legacy 🤔 it works as before. but I don't think all the new features work.

I corrected typos and added more stories exemple

shilman commented:
cc @mandarini

ThibaudAV committed:
feat(angular): componentWrapperDecorator accept props

Note : compatible with StoryContext (args, globals, ...)

ThibaudAV commented:
I add the option to pass props into `componentDecorator` and `templateDecorator`

and I open PR

shilman reviewed (approved):
Tremendous work. Bravo! 👏 👏 👏

ThibaudAV committed:
docs(angular): add snippets for `storybook-preview-with-styled-components-decorator`

ThibaudAV commented:
@shilman & @gaetanmaisse 
Last update : 
- rename and merge `componentDecorator` and `templateDecorator` to `componentWrapperDecorator`
- componentWrapperDecorator accepts functions to generate props from SotryContext (not only args) (see this use case : https://discord.com/channels/486522875931656193/490770949910691862/797010937923502102)
- add one new snippets

gaetanmaisse commented on app/angular/src/client/preview/angular-beta/utils/NgComponentAnalyzer.ts:
```diff
@@ -88,6 +88,20 @@ export const isDeclarable = (component: any): boolean => {
+    ? Reflect.getOwnPropertyDescriptor(component, decoratorKey).value
+    : component[decoratorKey];
+
+  return !!(decorators || []).find((d) => d instanceof Component);
```
Could be replaced by `(decorators || []).some((d) => d instanceof Component);`

gaetanmaisse commented on app/angular/src/client/preview/decorateStory.test.ts:
```diff
@@ -7,6 +7,77 @@ import { StoryFnAngularReturnType } from './types';
+            args: {
+              parentInput: 'Parent input',
+              grandparentInput: 'grandparent input',
+              parentOutput: expect.any(Function),
```
`parentOutput` should be `jest.fn` no?

ThibaudAV replied:
yes or simple void function

gaetanmaisse reviewed (approved):
2 minors comments, all your work is awesome 💪🏻

MickL commented:
I cant find `componentWrapperDecorator()` where is that imported from?

ThibaudAV commented:
@MickL 
Does it help you? :
https://github.com/storybookjs/storybook/blob/fcc75299fb052285d41c40c9454a4c2a8b873164/examples/angular-cli/src/stories/core/decorators/componentWrapperDecorator/decorators.stories.ts

lkashef commented:
Hey @ThibaudAV 

Just tried your PR and it works when template is supplied but when it's empty or removed it repeats the component itself.

WithContent vs WithoutContent

<img width="392" alt="Screen Shot 2021-02-23 at 10 02 03 PM" src="https://user-images.githubusercontent.com/15927737/108941286-d174f200-7622-11eb-95a3-49bdbd57b128.png">


<img width="388" alt="Screen Shot 2021-02-23 at 10 02 12 PM" src="https://user-images.githubusercontent.com/15927737/108941275-cde16b00-7622-11eb-9581-6c61dd3b28a5.png">

---

The code in case it helps:

```
import { Meta, componentWrapperDecorator } from "@storybook/angular";
import { FsDataPairComponent } from "../fs-data-pair.component";

export default {
  title: "Molecules/Data Pair",
  component: FsDataPairComponent,
  decorators: [
    componentWrapperDecorator(FsDataPairComponent),
  ],
} as Meta;

export const WithContent = (args) => ({
    props: args,
    template: 'The content',
  });
export const EmptyContent = (args) => ({
  props: args,
  template: '',
});

//fs-data-pair.stories.ts
```


```
<p>fs-data-pair works!</p>
<ng-content></ng-content>

//fs-data-pair.component.html
```

ThibaudAV commented:
yes it seems to me that another person also reproduced this bug. 
I haven't tested it myself but, I will do the fix when I have time 🙈 

I think it's a bad condition that rejects the empty template :not-sure: (and so the Story fallback on the component)

ThibaudAV commented:
@lkashef this fix should correct this problem ;)

lkashef commented:
Awesome! thanks @ThibaudAV 🙇

manuelmeister commented on app/angular/src/client/preview/decorateStory.ts:
```diff
@@ -0,0 +1,54 @@
+  const component = story.component ?? context.parameters.component;
+
+  if (!template && component) {
+    template = computesTemplateFromComponent(component, story.props, '');
```
Why can't we set the innerTemplate from outside?
